File: python/workflow/metrics.py
"""
Minimal metrics implementation with counters, gauges, and histograms.

This is intentionally simple but thread-safe and production-usable for small deployments.
"""

from __future__ import annotations

import threading
from dataclasses import dataclass, field
from typing import Dict, List


@dataclass
class Counter:
    name: str
    value: int = 0
    _lock: threading.Lock = field(default_factory=threading.Lock, init=False)

    def inc(self, amount: int = 1) -> None:
        with self._lock:
            self.value += amount


@dataclass
class Gauge:
    name: str
    value: float = 0.0
    _lock: threading.Lock = field(default_factory=threading.Lock, init=False)

    def set(self, value: float) -> None:
        with self._lock:
            self.value = value


@dataclass
class Histogram:
    name: str
    buckets: List[float]
    counts: List[int] = field(init=False)
    _lock: threading.Lock = field(default_factory=threading.Lock, init=False)

    def __post_init__(self) -> None:
        self.counts = [0 for _ in self.buckets]

    def observe(self, value: float) -> None:
        with self._lock:
            for idx, bucket in enumerate(self.buckets):
                if value <= bucket:
                    self.counts[idx] += 1
                    return
            # If value is above all buckets, count in the last bucket.
            self.counts[-1] += 1


class MetricsRegistry:
    """
    Thread-safe registry for metrics.
    """

    def __init__(self) -> None:
        self._counters: Dict[str, Counter] = {}
        self._gauges: Dict[str, Gauge] = {}
        self._histograms: Dict[str, Histogram] = {}
        self._lock = threading.Lock()

    def counter(self, name: str) -> Counter:
        with self._lock:
            if name not in self._counters:
                self._counters[name] = Counter(name=name)
            return self._counters[name]

    def gauge(self, name: str) -> Gauge:
        with self._lock:
            if name not in self._gauges:
                self._gauges[name] = Gauge(name=name)
            return self._gauges[name]

    def histogram(self, name: str, buckets: List[float]) -> Histogram:
        with self._lock:
            if name not in self._histograms:
                self._histograms[name] = Histogram(name=name, buckets=buckets)
            return self._histograms[name]

File: go/metrics/metrics.go
package metrics

import "sync"

// Counter is a thread-safe counter metric.
type Counter struct {
	name  string
	value int64
	mu    sync.Mutex
}

func (c *Counter) Inc(amount int64) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.value += amount
}

func (c *Counter) Value() int64 {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.value
}

// Gauge is a thread-safe gauge metric.
type Gauge struct {
	name  string
	value float64
	mu    sync.Mutex
}

func (g *Gauge) Set(value float64) {
	g.mu.Lock()
	defer g.mu.Unlock()
	g.value = value
}

func (g *Gauge) Value() float64 {
	g.mu.Lock()
	defer g.mu.Unlock()
	return g.value
}

// Histogram is a thread-safe histogram metric.
type Histogram struct {
	name    string
	buckets []float64
	counts  []int64
	mu      sync.Mutex
}

func NewHistogram(name string, buckets []float64) *Histogram {
	return &Histogram{name: name, buckets: buckets, counts: make([]int64, len(buckets))}
}

func (h *Histogram) Observe(value float64) {
	h.mu.Lock()
	defer h.mu.Unlock()
	for idx, bucket := range h.buckets {
		if value <= bucket {
			h.counts[idx]++
			return
		}
	}
	if len(h.counts) > 0 {
		h.counts[len(h.counts)-1]++
	}
}

// Registry stores metrics in a thread-safe map.
type Registry struct {
	mu         sync.Mutex
	counters   map[string]*Counter
	gauges     map[string]*Gauge
	histograms map[string]*Histogram
}

func NewRegistry() *Registry {
	return &Registry{
		counters:   make(map[string]*Counter),
		gauges:     make(map[string]*Gauge),
		histograms: make(map[string]*Histogram),
	}
}

func (r *Registry) Counter(name string) *Counter {
	r.mu.Lock()
	defer r.mu.Unlock()
	if c, ok := r.counters[name]; ok {
		return c
	}
	c := &Counter{name: name}
	r.counters[name] = c
	return c
}

func (r *Registry) Gauge(name string) *Gauge {
	r.mu.Lock()
	defer r.mu.Unlock()
	if g, ok := r.gauges[name]; ok {
		return g
	}
	g := &Gauge{name: name}
	r.gauges[name] = g
	return g
}

func (r *Registry) Histogram(name string, buckets []float64) *Histogram {
	r.mu.Lock()
	defer r.mu.Unlock()
	if h, ok := r.histograms[name]; ok {
		return h
	}
	h := NewHistogram(name, buckets)
	r.histograms[name] = h
	return h
}
